[Pytorch] Expose ops present in dispatcher

JacobSzwejbka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54791 [Pytorch] Expose ops present in dispatcher**

Several usecases have a need to want and see what ops are present in a specific pytorch runtime. This diff exposes that information in the dispatcher

Differential Revision: [D27271371](https://our.internmc.facebook.com/intern/diff/D27271371/)

Commit:
[Pytorch] Expose ops present in dispatcher

Several usecases have a need to want and see what ops are present in a specific pytorch runtime. This diff exposes that information in the dispatcher

Differential Revision: [D27271371](https://our.internmc.facebook.com/intern/diff/D27271371/)

[ghstack-poisoned]

Commit:
Update on "[Pytorch] Expose ops present in dispatcher"


Several usecases have a need to want and see what ops are present in a specific pytorch runtime. This diff exposes that information in the dispatcher

Differential Revision: [D27271371](https://our.internmc.facebook.com/intern/diff/D27271371/)

[ghstack-poisoned]

swolchok commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -138,6 +138,9 @@ class TORCH_API Dispatcher final {
   c10::optional<OperatorHandle> findOp(const OperatorName& operator_name);
 
+  // Returns a list of all operator names present in the operatorLookupTable_
+  const std::vector<OperatorName> getAllOpNames();
```
don't return const objects; it defeats RVO. See e.g. https://www.bfilipek.com/2017/01/const-move-and-rvo.html

swolchok commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -87,6 +87,16 @@ OperatorHandle Dispatcher::findSchemaOrThrow(const char* name, const char* overl
+const std::vector<OperatorName> Dispatcher::getAllOpNames() {
+  return operatorLookupTable_.read([&] (const ska::flat_hash_map<OperatorName, OperatorHandle>& operatorLookupTable) -> std::vector<OperatorName> {
+    std::vector<OperatorName> allOpNames;
+    for (const auto& op : operatorLookupTable) {
```
nit: reserve first -- `allOpNames.reserve(operatorLookupTable.size());`

swolchok reviewed (approved):
seems fine, please address comments before landing

Commit:
Update on "[Pytorch] Expose ops present in dispatcher"


Several usecases have a need to want and see what ops are present in a specific pytorch runtime. This diff exposes that information in the dispatcher

Differential Revision: [D27271371](https://our.internmc.facebook.com/intern/diff/D27271371/)

[ghstack-poisoned]
